doc/man/man4/sfx.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/sfx.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual under
.\" the terms of the GNU Free Documentation License, Version 1.3 or any later
.\" version published by the Free Software Foundation; with no Invariant
.\" Sections, no Front-Cover Texts, and no Back-Cover Texts.  A copy of the
.\" license is included in the section entitled "GNU Free Documentation
.\" License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaining to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF SUCH
.\" CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this manual
.\" page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from the
.\" use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
.\" behalf of the U.S. Government ("Government"), the following provisions
.\" apply to you.  If the Software is supplied by the Department of Defense
.\" ("DoD"), it is classified as "Commercial Computer Software" under
.\" paragraph 252.227-7014 of the DoD Supplement to the Federal Acquisition
.\" Regulations ("DFARS") (or any successor regulations) and the Government
.\" is acquiring only the license rights granted herein (the license rights
.\" customarily provided to non-Government users).  If the Software is
.\" supplied to any unit or agency of the Government other than DoD, it is
.\" classified as "Restricted Computer Software" and the Government's rights
.\" in the Software are defined in paragraph 52.227-19 of the Federal
.\" Acquisition Regulations ("FAR") (or any successor regulations) or, in the
.\" cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the
.\" FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR, \fB"
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH SFX 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Devices"
.\"
.\"
.SH NAME
.B sfx, fifo
\- \fISTREAMS\fR FIFO Device
.\"
.\"
.SH SYNOPSIS
.PP
none
.\"
.\"
.SH DESCRIPTION
.PP
The named pipe obtained by opening a special device node created with the
.BR mkfifo (1)
command are not necessarily \fISTREAMS\fR devices.  The
.B sfx
.I STREAMS
pseudo-device driver provides a mechanism for assigning
.IR STREAMS -based
fifos within the file system.
.PP
The
.B sfx
pseudo-device driver is a regular
.I STREAMS
character special device driver with the following exceptions:
.\"
.SS Characteristics
.PP
Following are some characteristics of FIFOs that cause their behaviour to differ from that of
regular STREAMS character special device files:
.IP \(bu \w'\(bu\(em'u
FIFOs cannot be clone opened.
.IP \(bu
FIFOs must be opened for read, write or both.  A FIFO cannot be opened otherwise.  (But see
.RI \(lq NOTICES \(rq,
below.)
.IP \(bu
FIFOs opened with the
.B O_NONBLOCK
flag set, will not block on the open, regardless of whether they are opened
.B O_RDONLY
and there are no writers, or opened
.B O_WRONLY
and there are no readers.
.IP \(bu
FIFOs opened with the
.B O_NONBLOCK
flag set, for write-only,
.BR O_WRONLY ,
operation, when no process has the FIFO open for reading will fail an return
.RB [ ENXIO ],
to the call to
.RB open (2s).
.IP \(bu
FIFOs opened with the
.B O_NONBLOCK
flag set, for read-only,
.BR O_RDONLY ,
operation, when no process has the FIFO open for writing will not block, will succeed and return
zero to the call to
.BR open (2s).
.IP \(bu
FIFOs opened without the
.B O_NONBLOCK
flag set, will block on the open if opened
.B O_RDONLY
and there are no writers, or opened
.B O_WRONLY
and there are no readers.
.IP \(bu
FIFOs opened without the
.B O_NONBLOCK
flag set, will not block on the open if opened
.BR O_RDWR .
.IP \(bu
When a FIFO is hung up, write and write-like operations return
.RB [ ENXIO ]
as is the case for regular
.I STREAMS
devices.
This is true for
.BR write (2s),
.BR writev (2s),
.BR putmsg (2s),
.BR putpmsg (2s),
and
.BR streamio (7)
input-output controls that write messages to the
.IR Stream ,
such as
.BR I_FDINSERT (7).
.IP \(bu
If there are no writers at the time that a read or read-like operation is performed, and the
.I Stream
is set for blocking operation, the read operation will block awaiting writers to the
.IR Stream .
.IP \(bu
If there are no readers at the time that a write or write-like operation is performed, and the
the write or write-like operations fail, generate a
.RI { SIGPIPE }
signal to the callers, and return
.RB [ EPIPE ],
regardless of whether the file descriptor is set for blocking or non-blocking operation.
.IP \(bu
By default, when an asynchronous write error occurs, and a write or write-like operation is
performed, FIFOs will generate
.RI { SIGPIPE }
to the caller in addition to returning the error from the asynchronous write error message.  This
behavior can be suppressed by clearing the
.B SNDPIPE
write option flag using the
.BR I_SWROPT (7)
input-output control.  This is the reverse of the default behavior for regular
.I STREAMS
character special files.
.IP \(bu
By default, attempting to write a zero-length data message to a FIFO with
.BR write (2s),
.BR writev (2s),
.BR putmsg (2s),
or
.BR putpmsg (2s),
will succeed, return zero, but no message will be generated.  Generation of the message can be
enabled by setting the
.B SNDZERO
write option flag using the
.BR I_SWROPT (7)
input-output control.  This is the reverse of the default behaviour for regular
.I STREAMS
character special files.
.IP \(bu
FIFOs break large writes down into independent data messages of
.RB { PIPE_BUF }
bytes or less.  Writes of less than or equal to
.RB { PIPE_BUF }
bytes are guaranteed atomic (i.e. they will not be interleaved with data from writes by other
processes).  Writes
of greater than
.RB { PIPE_BUF }
byes are not guaranteed atomic (i.e. may be interleaved with data from writes by other processes).
This is default behaviour of all
.I STREAMS
character special files, unless altered by the driver.
.IP \(bu
FIFOs that have no modules pushed, even when full and blocked on write, will not block on close as
described under
.BR I_SETCLTIME (7),
because there is no driver write queue.
.IP \(bu
.BR I_LIST (7)
will not report a count for a driver (FIFOs have no driver).
.IP \(bu
.BR I_LINK (7),
.BR I_PLINK (7),
.BR I_UNLINK (7),
and
.BR I_PUNLINK (7),
will always fail with error
.RB [ EINVAL ]
when issued to a file descriptor associated with a FIFO.
.IP \(bu
.BR I_SENDFD (7)
and
.BR I_RECVFD (7)
can succeed on a FIFO (whereas they always fail on a regular
.I STREAMS
character special file).
.\"
.\"
.SH NOTICES
.PP
Normally a
.I STREAMS
special file is associated with a device number.  For
.BR sfx ,
the
.I STREAMS
special file is associated with the inode in the file system in which the
character special node was created.  This acts identically to a FIFO created
with
.BR mkfifo (1),
with the exception that a
.BR stat (2)
on the device node will say that it is an
.B S_IFCHR
device node.  A
.BR fstat (2)
on a file descriptor opening this device node in the file system will say that
the file descriptor is of type
.BR S_IFIFO .
.PP
It may be necessary to push the
.BR pipemod (4)
module on one end of the FIFO to maintain the sense of
.BR M_FLUSH .
.PP
Under
.B @PACKAGE_TITLE@
is it possible to make all
.B Linux
system FIFOs
.IR STREAMS -based.
This is accomplished during configuration of the package and when the package kernel modules are
loaded.
.PP
Under
.B Linux
a FIFO cannot be opened without read or write.  This is because
.B Linux
defines the
.B O_RDONLY
flag to
.BR open (2s)
to be zero, despite
.I POSIX
.[
susv3
.]
recommendations to the contrary.
.PP
.RB { PIPE_BUF }
under
.B @PACKAGE_TITLE@
is set to the historical default of
.B 4096
bytes.
.PP
.RB { PIPE_BUF }
characteristics are accomplished by setting the
.I STREAMS
minimum packet size to zero
.RB ( 0 ),
the maximum packet size to
.RB ( 4096 ),
the high water mark to
.RB ( 5120 ),
and the low water mark to
.BR ( 1024 ).
These are the defaults for all
.I Stream heads
and can only be altered by a driver or module.
.PP
Some
.I STREAMS
implementations do not permit
.IR STREAMS -based
pipes or FIFOs to be linked beneath a multiplexing driver.
.B @PACKAGE_TITLE@
permits
.I STREAMS -based
pipes and FIFOs to be linked beneath a multiplexing driver.
.\"
.\"
.SH DEVICES
.PP
.TP
.BR /dev/fifo ", " /dev/sfx
The external filesystem
.I STREAMS
FIFO devices.
.TP
.BR /dev/fifo
An instance of an external filesystem
.I STREAMS
FIFO character device.  This FIFO device instance is provided for testing, primarily by the
.BR test-fifo (8)
conformance verification suite.
.TP
.BR /dev/streams/sfx ", " /devices/sfx
The
.BR specfs (5)
shadow special filesystem
.I STREAMS
fifo device.
.TP
.BI /dev/streams/fifo/ nnn
Instances of
.BR specfs (5)
shadow special filesystem
.I STREAMS
FIFO devices.
.\"
.\"
.SH "SEE ALSO"
.PP
.BR mkfifo (1),
.BR stat (2),
.BR fstat (2),
.BR pipe (2s),
.BR pipe (4),
.BR spx (4).
.\"
.\"
.SH BUGS
.PP
.B sfx
and
.B fifo
have no known bugs.
.\"
.\"
.SH COMPATIBILITY
.PP
.B sfx
is a character special node based replacement for fifos.  See
.BR mkfifo (1).
.\"
.\"
.SH CONFORMANCE
.PP
SVID,
.[
svid
.]
XID,
.[
xbd5
.]
SUSv2,
.[
susv2
.]
SUSv3.
.[
susv3
.]
Conformance is verified using the
.BR test-fifo (8)
verification test suite.
.\"
.\"
.SH HISTORY
.PP
.B fifos
first appeared in SVR 3,
.[
svr3
.]
but were not implemented using
.IR STREAMS .
.\"
.\"
.[
$LIST$
.]
.TI


NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/sfx.c
/*****************************************************************************

 @(#) File: src/drivers/sfx.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/sfx.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

#ifdef NEED_LINUX_AUTOCONF_H
#include NEED_LINUX_AUTOCONF_H
#endif
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>

#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#include "sys/config.h"
#include "src/kernel/strsched.h"
#include "src/kernel/strsad.h"	/* for autopush */
#include "src/modules/sth.h"
#include "src/drivers/fifo.h"	/* for fifo stuff */

#define SFX_DESCRIP	"SVR 4.2 STREAMS-based FIFOs Driver"
#define SFX_EXTRA	"Part of UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define SFX_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define SFX_REVISION	"OpenSS7 src/drivers/sfx.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define SFX_DEVICE	"SVR 4.2 MP STREAMS-based FIFOs"
#define SFX_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define SFX_LICENSE	"GPL"
#define SFX_BANNER	SFX_DESCRIP	"\n" \
			SFX_EXTRA	"\n" \
			SFX_COPYRIGHT	"\n" \
			SFX_REVISION	"\n" \
			SFX_DEVICE	"\n" \
			SFX_CONTACT	"\n"
#define SFX_SPLASH	SFX_DESCRIP	" - " \
			SFX_COPYRIGHT	"\n"

#ifdef CONFIG_STREAMS_SFX_MODULE
MODULE_AUTHOR(SFX_CONTACT);
MODULE_DESCRIPTION(SFX_DESCRIP);
MODULE_SUPPORTED_DEVICE(SFX_DEVICE);
MODULE_LICENSE(SFX_LICENSE);
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_SFX_MODULE */

#ifndef CONFIG_STREAMS_SFX_NAME
//#define CONFIG_STREAMS_SFX_NAME "sfx"
#error "CONFIG_STREAMS_SFX_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_SFX_MODID
//#define CONFIG_STREAMS_SFX_MODID 9
#error "CONFIG_STREAMS_SFX_MODID must be defined."
#endif
#ifndef CONFIG_STREAMS_SFX_MAJOR
//#define CONFIG_STREAMS_SFX_MAJOR 0
#error "CONFIG_STREAMS_SFX_MAJOR must be defined."
#endif

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-sfx");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_SFX_MODULE
static
#endif
modID_t modid = CONFIG_STREAMS_SFX_MODID;

#ifdef CONFIG_STREAMS_SFX_MODULE
#ifndef module_param
MODULE_PARM(modid, "h");
#else
module_param(modid, ushort, 0444);
#endif
MODULE_PARM_DESC(modid, "Module id number for STREAMS-based FIFOs (0 for allocation).");
#endif				/* CONFIG_STREAMS_SFX_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_SFX_MODID));
MODULE_ALIAS("streams-driver-sfx");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

#ifndef CONFIG_STREAMS_SFX_MODULE
static
#endif
major_t major = CONFIG_STREAMS_SFX_MAJOR;

#ifdef CONFIG_STREAMS_SFX_MODULE
#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for STREAMS-based FIFOs (0 for allocation).");
#endif				/* CONFIG_STREAMS_SFX_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_CLONE_MAJOR) "-" __stringify(CONFIG_STREAMS_SFX_MAJOR));
MODULE_ALIAS("/dev/sfx");
//MODULE_ALIAS("devname:sfx");
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_SFX_MAJOR));
MODULE_ALIAS("/dev/streams/sfx");
MODULE_ALIAS("/dev/streams/sfx/*");
MODULE_ALIAS("/dev/streams/clone/sfx");
#endif				/* MODULE_ALIAS */
#endif				/* MODULE */

static struct module_info sfx_minfo = {
	.mi_idnum = CONFIG_STREAMS_SFX_MODID,
	.mi_idname = CONFIG_STREAMS_SFX_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat sfx_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat sfx_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static struct qinit sfx_rqinit = {
	.qi_putp = strrput,
	.qi_srvp = strrsrv,
	.qi_qopen = str_open,
	.qi_qclose = str_close,
	.qi_minfo = &sfx_minfo,
	.qi_mstat = &sfx_rstat,
};

static struct qinit sfx_wqinit = {
	.qi_putp = strwput,
	.qi_srvp = strwsrv,
	.qi_minfo = &sfx_minfo,
	.qi_mstat = &sfx_wstat,
};

#ifdef CONFIG_STREAMS_SFX_MODULE
static
#endif
struct streamtab sfxinfo = {
	.st_rdinit = &sfx_rqinit,
	.st_wrinit = &sfx_wqinit,
};

static struct cdevsw sfx_cdev = {
	.d_name = CONFIG_STREAMS_SFX_NAME,
	.d_str = &sfxinfo,
	.d_flag = D_MP,
	.d_fop = &strm_f_ops,
	.d_mode = S_IFIFO | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

#ifdef CONFIG_STREAMS_SFX_MODULE
static
#endif
int __init
sfxinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_SFX_MODULE
	printk(KERN_INFO SFX_BANNER);
#else
	printk(KERN_INFO SFX_SPLASH);
#endif
	sfx_minfo.mi_idnum = modid;
	if ((err = register_strdev(&sfx_cdev, major)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_SFX_MODULE
static
#endif
void __exit
sfxexit(void)
{
	unregister_strdev(&sfx_cdev, major);
};

#ifdef CONFIG_STREAMS_SFX_MODULE
module_init(sfxinit);
module_exit(sfxexit);
#endif
